cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/worker/WorkerProgramRunner.java

```diff
@@ -110,6 +120,23 @@ public ProgramController run(Program program, ProgramOptions options) {
     return controller;
   }
 
+  @Nullable
+  private AdapterSpecification getAdapterSpecification(Arguments arguments) {
```

Comment:
1. No static method (mainly is not good, and partly because the AdapterSpecification is going to move into cdap-api).
2. The current ProgramRunner hierarachy doesn't have an abstract base class. When we need it in many program runner, we can abstract the program runner logic into some abstract base. I can put a TODO there (just don't wanna do it in the PR and don't have time to create a separate PR to refactor it properly).